tools/shell.cpp

```diff
@@ -0,0 +1,61 @@
+#include "slang.h"
+#include <histedit.h>
+#include <cstdio>
+
+using namespace std;
+using namespace slang;
+
+
+const char *prompt(EditLine *e) {
+    return " > ";
+}
+
+int main(int argc, char *argv[]) {
+    ScriptSession session(true);
+    EditLine *el;
+    History *cmdHistory;
+    HistEvent ev;
+
+    el = el_init(argv[0], stdin, stdout, stderr);
+    ASSERT(el);
+    el_set(el, EL_PROMPT, &prompt);
+
+    cmdHistory = history_init();
+    ASSERT(cmdHistory);
+    history(cmdHistory, &ev, H_SETSIZE, 1000);
+    el_set(el, EL_HIST, history, cmdHistory);
+
+    int charsRead;
+    const char *line;
+
+    /*
+    BumpAllocator alloc;
+    Diagnostics diagnostics;
+    SourceManager& srcMgr = SyntaxTree::getDefaultSourceManager();
+    Preprocessor preprocessor(srcMgr, alloc, diagnostics);
```

Comment:
yeah, the preprocessor seems tricky to get working here: I think the ideal thing would be to add the ability for the ScriptSession to maintain preprocessor stain between evaluations, otherwise this would probably be hard.